lean/2022-01-09-Graph-Coloring-Basics.lean

```diff
@@ -0,0 +1,139 @@
+/-
+layout: post
+title: "Graph Coloring Basics"
+date: 2022-01-09
+categories: graph
+
+Prelude:
+
+* [Graph homomorphism](https://en.wikipedia.org/wiki/Graph_homomorphism)
+* [NP-hardness](https://en.wikipedia.org/wiki/NP-hardness)
+
+# Graph Coloring Basics
+
+In Graph Theory, finding a proper coloring of a graph regards to attributing "colors"
+to its vertices in a way that adjacent vertices have different colors.
+
+[In Lean](https://github.com/leanprover-community/mathlib/pull/10287), a coloring (short
+for "proper coloring") is represented as a homomorphism from the graph being colored to
+the complete graph of available colors. Let's seen an example of how this works.
+
+Suppose we have a graph `G` that we want to color with three colors: `c1`, `c2` and `c3`.
+
+```
+      G            Complete graph of colors
+  D-------A                   c₁
+  |       | \                /  \
+  |       |  C              c₂---c₃
+  |       | /
+  E-------B
+```
+
+The following mapping can be seen as a coloring of `G`:
+
+```
+A → c₁
+B → c₂
+C → c₃
+D → c₂
+E → c₃
+```
+
+Now note that this mapping is in fact a homomorphism. And you may wonder, why does such
+a homomorphism represent a coloring?
+
+Informally, consider two adjacent vertices in G. Since our mapping is an homomorphism,
+those vertices must be mapped to adjacent vertices in the complete graph of colors.
+And such adjacent vertices represent different colors.
+
+You can check Mathlib's formalization of the above [here](https://leanprover-community.github.io/mathlib_docs/combinatorics/simple_graph/coloring.html#simple_graph.coloring.valid).
+
+Let's play with the API a bit.
+
+-/
+
+import combinatorics.simple_graph.coloring
+import tactic.derive_fintype
+
+@[derive [decidable_eq, fintype]]
+inductive verts : Type
+| A | B | C | D | E
+
+open verts
+
+def edges : list (verts × verts) :=
+[ (A, B), (B, C), (A, C), (A, D), (D, E), (B, E) ]
+
+/- This function tells whether two vertices are adjacent. -/
+
+def adj (v w : verts) : bool := edges.mem (v, w) || edges.mem (w, v)
+
+/- Coercing `bool` to `Prop`: -/
```

Comment:
This seems perhaps a bit unmotivated (though I may be able to tell what the reason is once I keep reading, presumably some piece of the API wants Props?)